proposals/0025-shader-execution-reordering.md

```diff
@@ -0,0 +1,1460 @@
+is treated as `write(caller)`
+
+### ReorderThread HLSL Additions
+
+`ReorderThread` provides an efficient way for the application to reorder work
+across the physical threads running on the GPU in order to improve the
+coherence and performance of subsequently executed code. The target ordering
+is given by the arguments passed to `ReorderThread`. For example, the
+application can pass a `HitObject`, indicating to the system that coherent
+execution is desired with respect to a ray hit location in the scene.
+Reordering based on a `HitObject` is particularly useful in situations with
+highly incoherent hits, e.g., in path tracing applications.
+
+`ReorderThread` is available only in shaders of type `raygeneration`.
+
+This function introduces a [Reorder Point](#reorder-points).
+
+#### Example 1
+
+The following example shows a common pattern of combining `HitObject` and
+`ReorderThread`:
+
+```C++
+// Trace a ray without invoking closesthit/miss shading.
+HitObject hit = HitObject::TraceRay( ... );
+
+// Reorder by hit point to increase coherence of subsequent shading.
+ReorderThread( hit );
+
+// Invoke shading.
+HitObject::Invoke( hit, ... );
+```
+
+---
+
+#### ReorderThread with HitObject
+
+This variant of `ReorderThread` reorders calling threads based on the
+information contained in a `HitObject`.
+
+It is implementation defined which `HitObject` properties are taken into
+account when defining the ordering. For example, an implementation may decide
+to order threads with respect to their hit group index, hit locations in
+3d-space, or other factors.
+
+`ReorderThread` may access both information about the instance in the
+acceleration structure as well as the shader record at the shader table
+offset contained in the `HitObject`. The respective fields in the `HitObject`
+must therefore represent valid instances and shader table offsets.
+NOP-HitObjects is an exception, which do not contain information about a hit
+or a miss, but are still legal inputs to `ReorderThread`. Similarly, a
+`HitObject` constructed from a `RayQuery` but did not set a shader table
+index is exempt from having a valid shader table record.
+
+```C++
+void ReorderThread( HitObject Hit );
+```
+
+Parameter                           | Definition
+---------                           | ----------
+`HitObject Hit` | `HitObject` that encapsulates the hit or miss according to which reordering should be performed.
+
+---
+
+#### ReorderThread with coherence hint
+
+This variant of `ReorderThread` reorders threads based on a generic
+user-provided hint. Similarity of hint values should indicate expected
+similarity of subsequent work being performed by threads. The resolution of
+the hint is implementation-specific. If an implementation cannot resolve all
+values of `CoherenceHint`, it is free to ignore an arbitrary number of least
+significant bits. The thread ordering resulting from this call may be
```

Comment:
We'd like to have a clearer description of how hit is mapping to coherence.
Could we add the sentence that higher bit of hint has more impact on coherence then lower ones.

e.g. consider 3 hints, and chose the most coherent pair of hint:
`A = 0b100000`
`B = 0b111111`
`C = 0b011111`

For example 3 methods for extracting coherence by defining opposite divergence metric:
1. If higher bit has more coherence then all lower bits after it, `A`,`B` pair is most coherent. Divergence more or less xor(...)
2. If its arithmetic difference then `A` and `C`. Divergence is more or less abs(substract(...)).
3. It could be also divergence = bitcount(xor) that tries to find most similar bits. then B and C is most coherent.

3 doesn't cope well with "it is free to ignore an arbitrary number of least
significant bits", but 1 and 2 fit well into definition.

I'm mentioning that because we discuss this feature only as reordering and forget that based on the ordering, the system needs to extract waves to execute.
Without understanding what is the actual "distance" between the hint codes its hard. It would be good to have that understanding on both sides of GPU.

Reply:
I believe arithmetic difference is the most flexible and intuitive definition for users. This approach allows a hint to incorporate both hierarchical aspects (with more important elements in MSBs) and ranges of bits that represent closeness along some axis. I agree it should be spelled out.

Reply:
Agreed that it makes sense to add some details what we mean with a hierarchical hint, and giving an example.

However, I think the numerical value of the XOR makes more sense than arithmetic difference. We could also mention lexicographical sorting (even if the implementation doesn't explicitly sort) which many users should be familiar with. 

Consider:

```
hint1 = 0b1000
hint2 = 0b0111
hint3 = 0b1111
```

If the bits in the hint are intended to be used hierarchically (as per my understanding), then `hint2` is closer to `hint1` than to `hint3`. That is reflected in the numerical value of the XOR: `hint2 ^ hint1 = 0b111`, while `hint2 ^ hint3 = 0b1000`. However, using arithmetic difference, `hint2` would be closer to `hint3`.

Reply:
XOR prioritizes the strictly hierarchical use case. The problem with XOR is when a bit range represents similarity along some linear axis and we want to group based on closeness. 

Take for example the following X,Y hint pairs along some linear axis of similarity:
```
X  Y  Arith XOR
5  0  5     5
5  1  4     4
5  2  3     7
5  3  2     6
5  4  1     1
5  5  0     0
5  6  1     3
5  7  2     2
5  8  3     13
5  9  4     12
```

In this use case, arithmetic difference produces a desirable metric but not XOR.

Reply:
I was interpreting the hints to have hierarchical semantics, but checking the spec again I couldn't find any explicit mentioning of that, so fine with me.

Again, good point that we should explain this.

Reply:
I view it like in hint one encodes closeness in two metrics:
1. discreet space, any difference on given bits mean you are different, on this and that property. e.g.
   - you will use resource with this and that indices
   - shader will enter some if for some particular effect
2. some numeric that is arithmetic based on which you can calculate distance. the only example comes to my mind,  as @rasmusnv, mentions some linear things. I think the example that will be used most likely is parts of morton coded directions and origins.

now, the problem with arithmetic is that it potentially see the two hints with all bits different as the closest, so it breaks the 1. ... totally

on the other hand xor is succesfully used for checking morton builder.
even more crude, the closer code is considered the one that has lower first different bit (maybe I should put that on a list two, actually). so xor IMO should work great for 1. and good for 2. IMHO.

Reply:
I agree that there are some interesting aspects to specifying XOR between the two, but I think we need to consider something like this in the context of the entire system to see if it's reasonable.

Let's say a sort is performed that will put the hint values in descending order. Chunks of this sorted work is distributed to various execution units for dispatch. This should work for both arithmetic distance (it's an arithmetic sort) and XOR distance (the MSB has the highest weight on its order). The possible approximation with respect to the distance metric is at chunk boundaries.

As mentioned, this still leaves the question of how to prioritize dispatch groups within each execution unit when values are not a perfect match. This assumes that we are doing something sophisticated, rather than simply grabbing the next N sorted items.

In the earlier example, the three values would be ordered as:

```
B = 0b111111
A = 0b100000
C = 0b011111
```

Let's assume that we only have to group 2 items. We either pick A&B or A&C depending on if it's a XOR or arithmetic distance metric. Another choice is a strict **order** metric, which is actually the simplest in this case. That would indicate that we start with A&B.

Then consider the case where we have to group more work, say 8. Let's say that we have some duplicates (x-factor indicates the number of duplicates):

B = 0b111111 x 4
A = 0b100000 x 1
C = 0b011111 x 4

Would we be better served grouping B&C, B&A, A&C? I think this is not clear. What I'm trying to say is that the exact distance metric for dispatch grouping might be best left unspecified.

Maybe a suitable formulation would be that the system should in general try to dispatch hint values in arithmetic order (NOTE: not distance), but the exact mapping to waves is left unspecified.

Thoughts?

Reply:
I would prefer if the expectation was we'd group A&B in your second example, based on the `XOR` distance. Given the grouping is allowed to throw away lower-order bits, it already implies higher order bits are more important, and it will be likely used to set "feature bits" anyways by ISVs because "arithmetic distance" between shaders is not really something you can calculate. But you can say "those two shaders are going to do X" and we want to group by similarity, so the more similar bits are set the more likely they are similar, with higher bits used to indicate "more expensive" things.

Undefined means it'll be defined by implementations, after all.

Reply:
One thing to note here is that the XOR codes can be converted to arithmetic codes while preserving distance orderings at the cost of using a few more bits.  One such mapping is shown below.

![image](https://github.com/user-attachments/assets/cefe2da5-a2d1-4af4-aebc-91839893ea89)

In the a second example above we would have:

Label | XOR code | Distance preserving arithmetic code
---|---|---
B | 63 | 364
A | 32 | 243
C | 31 | 121

With this transform we can use an arithmetic distance, yet B&A remain closer than A&C.

Reply:
I'm not convinced that it's easier for _applications_ to produce an arithmetic distance given applications will know features, not distances. I.e. a bit corresponds to a particular property, it's hard to imagine how an application would calculate an arithmetic distance from a set of feature bits. To me we got this backwards - now you'd have to take a "feature based metric" from the app, convert it to arithmetic distance using the reverse encoding, then unpack it to XOR again for sorting - what did we win? Would you mind posting the formula for both directions in this case?

Reply:
For clarity, in the above the application isn't computing an arithmetic _distance_.  It's converting an xor _code_ into an arithmetic _code_ directly, such that the distance orderings are preserved.

The code conversion above reinterprets the xor code bits in trinary.  So where a bit would normally be interpreted to have value 2^i the conversion interprets it as 3^i.  So xor code 0b0101 (aka 5) becomes 3^2 + 3^0 = 10.  This inserts a gap between each new option bit sufficient to preserve distance orderings.

In application code this could be a change from `xor_code |= option_bit;` to `arith_code += option_value;`.  Of course we could imagine more complex scenarios than simply accumulating option constants.

I don't think it is possible to preserve arithmetic distance orderings using an xor code.  Maybe it can be done but it's not obvious to me how to do it.  So unfortunately I can't post the formula for the opposite direction.

>take a "feature based metric" ... what did we win?

Since there appear to be valid arguments for both arithmetic and xor distances what we won is having a single consistent code space (arithmetic) which the application can use to encode either arithmetic or xor distances.  The implementations can always sort and group by arithmetic distance and the application has the freedom to construct its codes to reflect either distance ordering.

Reply:
The most important thing about arithmetic is that for me its not even good for arithmetic based HW if it drops bits.
So impl has to do rounding when doing dropping.

e.g.
original codes given by app:
A 101000000
B 100111111
C 101111111
if implementation drops 6 bits it sees:
A 101__________
B 100__________
C 101__________

For arithmetic the closest codes changes from A and B to A and C when bits are dropped.
for XOR it stays A and C.
Simply arithmetic is not stable for dropping of bits.


I wonder if we could pass literal with mask that says at least how the structure of hint looks like. for instance if hint by app is of structure:

```c
struct HintStructA {
  uchar featureA : 1;
  uchar featureB : 1;
  uchar LittleEnum : 2;
  uchar featureD : 1;
  uchar LittleArithCode : 3;
};
```
than ReorderThreads instead of having important bit count argument could have mask (compile time constant!) that marks on which bit each field starts. 
For above `hintStructA`: it would be `0b11101100`;

Dropping bits impl may be dropping whole fields. App can express its wish of understanding the hint including the ones we are discussing above:
e.g. 
arithmetic 8bit 0b10000000
xor            8bit 0b11111111